include/pmacc/particles/memory/dataTypes/SuperCell.hpp

```diff
@@ -22,70 +22,93 @@
 #pragma once
 
 #include "pmacc/types.hpp"
+#include "pmacc/math/vector/compile-time/Vector.hpp"
 
-namespace pmacc
-{
 
-template <class TYPE>
-class SuperCell
+namespace pmacc
 {
-public:
-
-    HDINLINE SuperCell() :
-    firstFramePtr(nullptr),
-    lastFramePtr(nullptr),
-    mustShiftVal(false),
-    sizeLastFrame(0)
-    {
-    }
-
-    HDINLINE TYPE* FirstFramePtr()
-    {
-        return firstFramePtr;
-    }
-
-    HDINLINE TYPE* LastFramePtr()
-    {
-        return lastFramePtr;
-    }
-
-    HDINLINE const TYPE* FirstFramePtr() const
-    {
-        return firstFramePtr;
-    }
 
-    HDINLINE const TYPE* LastFramePtr() const
+    template< class T_FrameType >
+    class SuperCell
     {
-        return lastFramePtr;
-    }
-
-    HDINLINE bool mustShift()
-    {
-        return mustShiftVal;
-    }
-
-    HDINLINE void setMustShift(bool value)
-    {
-        mustShiftVal = value;
-    }
-
-    HDINLINE lcellId_t getSizeLastFrame()
-    {
-        return sizeLastFrame;
-    }
-
-    HDINLINE void setSizeLastFrame(lcellId_t size)
-    {
-        sizeLastFrame = size;
-    }
-
-
-private:
-    PMACC_ALIGN(mustShiftVal, bool);
-    PMACC_ALIGN(sizeLastFrame, lcellId_t);
-public:
-    PMACC_ALIGN(firstFramePtr, TYPE*);
-    PMACC_ALIGN(lastFramePtr, TYPE*);
-};
-
-} //end namespace
+    public:
+
+        HDINLINE SuperCell() :
+            firstFramePtr( nullptr ),
+            lastFramePtr( nullptr ),
+            numParticles( 0 ),
+            mustShiftVal( false )
+        {
+        }
+
+        HDINLINE T_FrameType * FirstFramePtr()
+        {
+            return firstFramePtr;
+        }
+
+        HDINLINE T_FrameType * LastFramePtr()
+        {
+            return lastFramePtr;
+        }
+
+        HDINLINE T_FrameType const * FirstFramePtr() const
+        {
+            return firstFramePtr;
+        }
+
+        HDINLINE T_FrameType const*  LastFramePtr() const
+        {
+            return lastFramePtr;
+        }
+
+        HDINLINE bool mustShift() const
+        {
+            return mustShiftVal;
+        }
+
+        HDINLINE void setMustShift( bool const value )
+        {
+            mustShiftVal = value;
+        }
+
+        HDINLINE uint32_t getSizeLastFrame() const
+        {
+            constexpr uint32_t frameSize = math::CT::volume<
+                typename T_FrameType::SuperCellSize
+            >::type::value;
+            uint32_t numParLastFrame = numParticles % frameSize;
```

Comment:
Typo: Par instead of Per

Reply:
Ahh it take me a few seconds to get you comment, `Par` means particles. I will change it to `numParticlesInLastFrame`

Reply:
Ah, sorry, I thought it was meant to be "number per last frame".